socorro/app/socorro_app.py

```diff
@@ -202,13 +207,18 @@ def _do_run(klass, config_path=None, values_source_list=None):
                 # pull in the application defaults from the 'application'
                 # configman option, once it has been defined
                 application_defaults_proxy,
+                # Get values from the app's config module
+                klass.config_module,
```

Comment:
So here's the problem line. Since the processor is run directly by docker, `klass` is set to `ProcessorApp` and this finds the correct value. When an app is run via `SocorroWelcomeApp`, that's the value of `klass` here, because the welcome app has no idea what application it is running until after the config is generated.

This is why `for_application_defaults.py` has that proxy stuff, it uses `from_str_converter` to be notified when the application was selected, and then subs in values properly after that point. I did not fully understand why it was doing that until now.

Honestly, I think our best bet is ditching `SocorroWelcomeApp` completely and running all apps directly in the same way the processor is. It doens't seem like a useful level of indirection at this point, but I may be missing something.

The second best solution is probably to store the config module path in the `SOCORRO_APPS` constant so we know it ahead of time. The problem with this is that it makes it a bit awkward to do config inheritance, which mainly affects subclasses of `FetchTransformSaveApp`, which has some default config values it declares. However, I'm not sure if anyone is even using those default values, so this might actually work just fine too.

@willkg Thoughts? Can we kill `SocorroWelcomeApp`?

Reply:
`SocorroWelcomeApp` is how `./scripts/socorro <appname>` works. In a different conversation we tossed around a way of coordinating scripts where they're subcommands of a single command. I like that model because I think it'll make discovery and documentation easier.

Would it be possible to rewrite `SocorroWelcomeApp` so it doesn't do any configman things, but instead just plucks off the subcommand, matches that to the app via a lookup somewhere, and then kicks off that app? I'd be game for taking a stab at that in a separate PR after this one lands.

Reply:
It's possible. I was thinking of using a library like [click](http://click.pocoo.org/) for the central script runner, but I dunno if that would conflict with the commandline features of configman. I also dunno if we want this command to be both a tool you'd use in the local development environment, and in the server environments to start off various apps or not.

I'll remove the welcome app as part of this PR and convert all the uses in-repo of it to just call the app directly (I believe they're almost all for crontabber), and then we can figure out what to do about a single command after.

Reply:
I don't think we need to add click to do subcommands. I think we can change `scripts/socorro` to something like this:

```
import sys

# app_map would be a dict mapping app name to python dotted path
from socorro.app.app_registry import app_map
from socorro.app.app_registry import str_to_application_class

USAGE = 'socorro <subcommand> [options]'


def print_usage(msg=''):
    print(USAGE)
    print('Available subcommands:')
    for cmd in sorted(app_map.keys()):
        print('   ' + cmd)


def main(args):
    if not args or args[1] == '--help':
        print_usage()
        return 0

    if args[0] not in app_map:
        print('%s is not a valid subcommand. See "socorro --help".' % args[0])
        return 0

    # Knock the subcommand off the front of the args list
    sys.argv = sys.argv[1:]

    # Create the App class and then run it
    app_class = str_to_application_class(app_map[args[0]])
    return app_class.run(config_path=None, values_sources_list=None)


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
```

Then `app_registry.py` has a map of app names to dotted Python paths just like `for_application_defaults` has (sort of).

Or something really close to that.

Then the following would all work and do helpful things:

```
# Prints list of commands
$ ./scripts/socorro
$ ./scripts/socorro --help

# Runs command help
$ ./scripts/socorro crontabber --help

# Runs a command
$ ./scripts/socorro crontabber ...
```

click is pretty super, but I don't think we need it for this case.

I think your plan sounds good and we can redo the socorro script or whatever we decide in a follow-up PR.